### PR TITLE
Fix rockdb iterator leak for local cache

### PR DESCRIPTION
**Pull request by beinan**
### What changes are proposed in this pull request?

Fix rockdb iterator leak for local cache by using the new rocksdb iterator pattern in RocksUtil

### Why are the changes needed?

As @dbw9580 suggested, it might cause the segment fault to use the raw rocksdb iterator

### Does this PR introduce any user facing changes?

no

### CODE REVIEW

**Comment by beinan** [comment]
@jiacheliu3 and @dbw9580 , could you help take a look of this PR? Thanks!

File: core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java [hunk 1]
```diff
@@ -79,7 +79,7 @@ public static long readLong(byte[] bytes, int start) {
    *
    * @param <T> return type of parser's next method
    */
-  interface RocksIteratorParser<T> {
+  public interface RocksIteratorParser<T> {
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java:82]: is this used?

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java [hunk 2]
```diff
@@ -99,22 +116,21 @@ private static RocksDB openDB(RocksPageStoreOptions pageStoreOptions, Options ro
    *
    * @param pageStoreOptions options for the rocks page store
    * @param rocksDB RocksDB instance
-   * @param dbOptions the RocksDB options
    */
-  private RocksPageStore(RocksPageStoreOptions pageStoreOptions, RocksDB rocksDB,
-      Options dbOptions) {
-    mPageStoreOptions = pageStoreOptions;
-    mDbOptions = dbOptions;
+  private RocksPageStore(RocksPageStoreOptions pageStoreOptions,
```
Comment (jiacheliu3) [inline core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java:120]: One great thing about RocksDB JNI is, for everything that extends `RocksObject` you will need to call `close()` on it, for example, double check your `ColumnFamilyHandle`.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java [hunk 3]
```diff
@@ -55,32 +63,41 @@ public static RocksPageStore open(RocksPageStoreOptions pageStoreOptions) {
     Preconditions.checkArgument(pageStoreOptions.getMaxPageSize() > 0);
     RocksDB.loadLibrary();
     // The RocksObject will be closed together with the RocksPageStore
-    Options rocksOptions = new Options()
+    DBOptions rocksOptions = new DBOptions()
```
Comment (jiacheliu3) [inline core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java:66]: also check if this is closed

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
My experience is, whenever i see an import from rocksDB, i will manually check if that extends `RocksObject` and if so, it needs to be manually closed in code.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java (hunk 3 above)
Comment (dbw9580) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java:66]: From the javadoc on `RocksDB.open`:

> Options instance *should* not be disposed before all DBs using this options instance have been closed. If user doesn't call options dispose explicitly, then this options instance will be GC'd automatically

Make sure to keep the reference to the options object live, and close it until the DB is closed.

File: core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java (hunk 1 above)
Comment (beinan) [reply core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java:82]: this is not used explicitly, but the compiler will complain on the call of `RocksUtils.createCloseableIterator` if it's not public

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java (hunk 2 above)
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java:125]: Thank you for this great tricks!  I just explicitly called close for every RocksObject I can find.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java (hunk 3 above)
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/RocksPageStore.java:68]: Yep, as @dbw9580 suggested, I just close the DBOptions on the db close.  Thank you!

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thanks! Good to merge once the tests are fixed.

**Comment by beinan** [comment]
alluxio-bot, merge this please.

**Pull request merged**